packages/@glimmer/runtime/lib/compiled/opcodes/expressions.ts

```diff
@@ -395,3 +395,49 @@ APPEND_OPCODES.add(Op.GreaterEqual, (vm) => {
     })
   );
 });
+
+APPEND_OPCODES.add(Op.And, (vm) => {
+  let { positional } = check(vm.stack.popJs(), CheckArguments).capture();
+  let stack = vm.stack;
+
+  stack.pushJs(
+    createComputeRef(() => {
+      let values = reifyPositional(positional);
+
+      // this is handlebars definition of truthiness
+      for (let i = 0, len = values.length; i < len; i++) {
+        if (isTruthy(values[i]) === false) {
+          return values[i];
+        }
+      }
+      return values[values.length - 1];
+    })
+  );
+});
+
+APPEND_OPCODES.add(Op.Or, (vm) => {
+  let { positional } = check(vm.stack.popJs(), CheckArguments).capture();
+  let stack = vm.stack;
+
+  stack.pushJs(
+    createComputeRef(() => {
+      let values = reifyPositional(positional);
+
+      // this is handlebars definition of truthiness
+      for (let i = 0, len = values.length; i < len; i++) {
+        if (isTruthy(values[i]) === true) {
+          return values[i];
+        }
+      }
+      return values[values.length - 1];
+    })
+  );
+});
+
+function isTruthy(result: unknown) {
```

Comment:
https://github.com/snewcomer/glimmer-vm/pull/2#discussion_r609861506

Reply:
Comments to respond to

> I guess for now I can just say them here. Suppose someone had the following:
> 
> {{or someMaybeEmptyArray otherArray}}
> If we're trying to do TS types for this we can't just convert to someMaybeEmptyArray || otherArray. As long as someMaybeEmptyArray isn't null or undefined TS will see it as being truthy and will just return the type as someMaybeEmptyArray however, the correct type would actually be someMaybeEmptyArray | otherArray. In the case where someone wants to handle empty arrays, I think it would more correctly done as:
> 
> {{if someMaybeEmptyArray.length someMaybeEmptyArray otherArray}}
> We could then correctly type that as someMaybeEmptyArray | otherArray.

Reply:
@snewcomer @wagenet

I also think that adopting "handlebars truthiness" for these helpers was a design mistake, and further that having handlebars truthiness differ from Javascript truthiness was itself a design mistake, though probably far too late to fix at this date.

That being said, does the typing issue actually need to be addressed internally in the glimmer VM?

Consumers will be getting their templates typechecked by glint, for which I am preparing a PR to add support for logical and equality helpers.

I think that the type for isTruthy (as glint would need to supply in its temporary translated TypeScript code) is something along the lines of:

```typescript
function isTruthy(v: true): true;
function isTruthy(v: false): false;
function isTruthy<A>(v: A[]): v is  [A, ...A[]];
function isTruthy<A>(v: readonly A[]): v is readonly [A, ...A[]];
function isTruthy(v: unknown): v is true | string | number | object;
function isTruthy(v: unknown): boolean {
  return !!v && !isEmptyArray(v);
}
```

It seems that the TypeScript compiler (as of 4.8.4) is not smart enough to propagate type assertions through a function of this type, however, so to support discriminated unions, glint needs to translate an expression of the form

```handlebars
{{or a b c}}
```
into
```typescript
(a && isTruthy(a)) ? a : (b && isTruthy(b)) ? b : c
```

which means that you can write things of the form:

```typescript
type X =
  | { type: 'a'; value: number }
  | { type: 'b'; value: number }
  | { type: 'c'; error: Error };

let foo: X;
```

```handlebars
{{#if (or (eq foo.type 'a') (eq foo.type 'b'))}}
  {{log foo.value}}
{{else}}
  {{log foo.error.message}}
{{/if}
```

Reply:
Actually, I'm not sure that the TypeScript compiler is quite smart enough to treat something like `(foo.type === 'a' && isTruthy(foo.type === 'a')) ? a : false` as equivalent to `foo.type === 'a'` for type-discrimination purposes but there is probably a way to translate the expressions that adequately expresses the semantics.

Reply:
I missed this discussion at first, but it's definitely the crux of the issue. If we can't generate useful types from this the benefit is significantly reduced. I'll see if I can poke at this a bit to figure out what'a possible here. No RFC is ever set in stone, but at the same time, diverging from ember-truth-helpers isn't ideal either.

Reply:
I'm not sure there's a way to capture the HBS semantics in a way that TS will propagate guards while _also_ making the fallthrough typing `{{or a b}}` correctly give `typeof a | typeof b` as its type without polluting that type with `| false` or similar.

I think even if this lands with the traditional HBS truthiness semantics, I'd still argue for treating them for typing purposes as though they were the native TS operators. I'm not set in stone on that (and I think there's a decent change @chriskrycho will feel differently 😄), but to me the gains from having logical and comparison operators "just work" for type narrowing outweighs the hazard of us typing `{{or stringArray numberArray}}` incorrectly as `Array<string>`.

Reply:
I'm putting this on the agenda for the Framework Core team meeting. The *whole point* of RFC Stages is that this kind of thing is useful to feed back in. I don’t want to reopen/relitigate the whole thing (YIKES) but I do want to make sure we are making a conscious and explicit choice to rule out having the ability for TS to provide both useful *and* accurate semantics for these.

Reply:
Having said all that, if `{{and a b c}}` is translated for typechecking purposes as `a && b && c`, then at least we can use it for most type guards.

With the present Handlbars truthiness semantics, the type of the expression `{{get (or a b c) 0}}` where `a` and `b` are `Array<T>` and `c` is a `NonEmptyArray<T>` would ideally be `T`, not `T | undefined`, but this is (hopefully?) going to rarely be expressed in a template, anyway, and if you really had to write that, you'd use `{{if}}` and an `is-empty` helper that provides a type guard.

Is there a case where `a && b && c` or `a || b || c` produces an unsound type when that the helper uses Handlebars truthiness, i.e. the code passes typechecking but will fail at runtime? Having thought about this for a bit, I'm not sure that can actually happen.

Reply:
If I have `strings: string[]` and `numbers: number[]`, then:

```hbs
{{#each (or strings numbers) as |el|}}
  {{! `el` is `string` here according to TS's `||`, but could be `number` if `strings` is empty given HBS truthiness }}
{{/each}}
```